File: go/0047M-全排列II.go
/*
 * @Author: SukiEva
 * @Date: 2021-11-15 09:24:05
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
package main

import "sort"

func permuteUnique(nums []int) [][]int {
	ans := make([][]int, 0)
	per := make([]int, 0)
	vis := make([]bool, len(nums))
	var dfs func(index int)
	dfs = func(index int) {
		if len(per) == len(nums) {
			tmp := make([]int, len(per))
			copy(tmp, per)
			ans = append(ans, tmp)
			return
		}
		for i := 0; i < len(nums); i++ {
			/*
				vis[i] = true 同一树枝使用过
				vis[i] = false 同一树层使用过
			*/
			if vis[i] {
				continue
			}
			if i > 0 && nums[i] == nums[i-1] && vis[i-1] == false {
				continue
			}
			per = append(per, nums[i])
			vis[i] = true
			dfs(i + 1)
			vis[i] = false
			per = per[:len(per)-1]
		}
	}
	sort.Ints(nums)
	dfs(0)
	return ans
}

File: redo/0047M-全排列II.cpp
/*
 * @Author: SukiEva
 * @Date: 2022-02-07 20:50:16
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   private:
    vector<vector<int>> ans;
    vector<int> tmp;
    bool vis[10] = {false};

    void dfs(vector<int>& nums) {
        if (tmp.size() == nums.size()) {
            ans.push_back(tmp);
            return;
        }
        for (int i = 0; i < nums.size(); i++) {
            if (vis[i]) continue;
            if (i > 0 && nums[i] == nums[i - 1] && !vis[i - 1]) continue;
            tmp.push_back(nums[i]);
            vis[i] = true;
            dfs(nums);
            tmp.pop_back();
            vis[i] = false;
        }
    }

   public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        dfs(nums);
        return ans;
    }
};